cmd/soroban-rpc/internal/db/ledger.go

```diff
@@ -31,10 +32,92 @@ type ledgerReader struct {
 	db *DB
 }
 
+type LedgerReaderTx interface {
+	GetLedgerRange(ctx context.Context) (ledgerbucketwindow.LedgerRange, error)
+}
+
+type ledgerReaderTx struct {
+	db                         *DB
+	latestLedgerSeqCache       uint32
+	latestLedgerCloseTimeCache int64
+}
+
+// GetLedgerRange pulls the min/max ledger sequence numbers from the meta table.
+func (r *ledgerReaderTx) GetLedgerRange(ctx context.Context) (ledgerbucketwindow.LedgerRange, error) {
+	// Make use of the cached latest ledger seq and close time to query only the oldest ledger details.
+	if r.latestLedgerSeqCache != 0 {
+		query := sq.Select("meta").
+			From(ledgerCloseMetaTableName).
+			Where(
+				fmt.Sprintf("sequence = (SELECT MIN(sequence) FROM %s)", ledgerCloseMetaTableName),
+			)
+		var lcm []xdr.LedgerCloseMeta
+		if err := r.db.Select(ctx, &lcm, query); err != nil {
+			return ledgerbucketwindow.LedgerRange{}, fmt.Errorf("couldn't query ledger range: %w", err)
+		}
+
+		return ledgerbucketwindow.LedgerRange{
+			FirstLedger: ledgerbucketwindow.LedgerInfo{
+				Sequence:  lcm[0].LedgerSequence(),
+				CloseTime: lcm[0].LedgerCloseTime(),
+			},
+			LastLedger: ledgerbucketwindow.LedgerInfo{
+				Sequence:  r.latestLedgerSeqCache,
+				CloseTime: r.latestLedgerCloseTimeCache,
+			},
+		}, nil
+	}
+
+	query := sq.Select("lcm.meta").
+		From(ledgerCloseMetaTableName + " as lcm").
+		Where(sq.Or{
+			sq.Expr("lcm.sequence = (?)", sq.Select("MIN(sequence)").From(ledgerCloseMetaTableName)),
+			sq.Expr("lcm.sequence = (?)", sq.Select("MAX(sequence)").From(ledgerCloseMetaTableName)),
+		}).OrderBy("lcm.sequence ASC")
+
+	var lcms []xdr.LedgerCloseMeta
+	if err := r.db.Select(ctx, &lcms, query); err != nil {
+		return ledgerbucketwindow.LedgerRange{}, fmt.Errorf("couldn't query ledger range: %w", err)
+	}
+
+	// Empty DB
+	if len(lcms) == 0 {
+		return ledgerbucketwindow.LedgerRange{}, ErrEmptyDB
+	}
+
+	r.latestLedgerSeqCache = lcms[len(lcms)-1].LedgerSequence()
+	r.latestLedgerCloseTimeCache = lcms[len(lcms)-1].LedgerCloseTime()
+
+	return ledgerbucketwindow.LedgerRange{
+		FirstLedger: ledgerbucketwindow.LedgerInfo{
+			Sequence:  lcms[0].LedgerSequence(),
+			CloseTime: lcms[0].LedgerCloseTime(),
+		},
+		LastLedger: ledgerbucketwindow.LedgerInfo{
+			Sequence:  lcms[len(lcms)-1].LedgerSequence(),
+			CloseTime: lcms[len(lcms)-1].LedgerCloseTime(),
+		},
+	}, nil
+}
+
 func NewLedgerReader(db *DB) LedgerReader {
 	return ledgerReader{db: db}
 }
 
+func (r ledgerReader) NewTx(ctx context.Context) (LedgerReaderTx, error) {
+	txSession := r.db.Clone()
+	if err := txSession.BeginTx(ctx, &sql.TxOptions{ReadOnly: true}); err != nil {
```

Comment:
this transaction is never closed and it's actually not passed into `ledgerReaderTx{}` .

I think a better approach would be to remove the `NewTx()` function and keep the `LedgerReader ` interface as it was before with `GetLedgerRange(ctx context.Context) (ledgerbucketwindow.LedgerRange, error)`.

`GetLedgerRange()` can encapsulate the transaction management

Reply:
I am going to remove the transaction creation as discussed with @tamirms on a call.